lib/Runtime/Base/FunctionExecutionStateMachine.cpp

```diff
@@ -376,147 +482,91 @@ namespace Js
-            if (GetInterpretedCount() < profilingInterpreterLimit)
-            {
-                VerifyExecutionMode(GetExecutionMode());
-                return false;
+                isTransitionNeeded = true;
             }
-            CommitExecutedIterations(profilingInterpreterLimit, profilingInterpreterLimit);
-            // fall through
-        }
 
-    TransitionFromProfilingInterpreter:
-        Assert(profilingInterpreter0Limit == 0 || profilingInterpreter1Limit == 0);
-        if (autoProfilingInterpreter1Limit == 0 && simpleJitLimit == 0 && profilingInterpreter1Limit == 0)
-        {
-            goto TransitionToFullJit;
-        }
-        goto TransitionToAutoProfilingInterpreter;
-
-        // Managing transition to and from SimpleJit
-    TransitionToSimpleJit:
-        if (simpleJitLimit != 0)
-        {
-            SetExecutionMode(ExecutionMode::SimpleJit);
-
-            // Zero the interpreted count here too, so that we can determine how many interpreter iterations ran
-            // while waiting for simple JIT
-            SetInterpretedCount(0);
-            return true;
-        }
-        goto TransitionToProfilingInterpreter;
-
-        case ExecutionMode::SimpleJit:
-        {
-            FunctionEntryPointInfo *const simpleJitEntryPointInfo = owner->GetSimpleJitEntryPointInfo();
-            if (!simpleJitEntryPointInfo || simpleJitEntryPointInfo->callsCount != 0)
+            if (isTransitionNeeded)
             {
-                VerifyExecutionMode(GetExecutionMode());
-                return false;
-            }
-            CommitExecutedIterations(simpleJitLimit, simpleJitLimit);
-            goto TransitionToProfilingInterpreter;
-        }
+                // Keep advancing the state until a terminal state is found or until there are no more 
+                // states to reach. The path of advancement is described in the banner comment above.
+                ExecutionState newState = executionState;
+                while (isTransitionNeeded && !IsTerminalState(newState))
+                {
+                    if (newState != ExecutionState::Interpreter)
+                    {
+                        // Most states simply advance to the next state
+                        newState = ExecutionState(newState + 1);
+                    }
+                    else
+                    {
+                        // Interpreter advances straight to FullJit
+                        newState = ExecutionState::FullJit;
+                    }
+
+                    if (newState != ExecutionState::FullJit || !PHASE_OFF(FullJitPhase, owner))
+                    {
+                        isTransitionNeeded = true;
+                    }
+                    else
+                    {
+                        // If FullJit is the next state, but FullJit is disabled, then no transition
+                        // is needed.
+                        isTransitionNeeded = false;
+                    }
```

Comment:
To me, this would be more readable if the conditional logic was flipped and isTransitionNeeded not changed if the (flipped) condition failed.

Reply:
actually--I can do better and simply set the bool to the conditional statement. not sure why I didn't think of that before, but i'll do it now :-)

Reply:
nope, I change my mind--I like your idea better. will be in the next update